### PR TITLE
Image Editor: clicking on "Done" performs cancel fn if image is not loaded

### PR DESCRIPTION
**Pull request by lamosty**
If image has not been loaded yet in the image editor, clicking on the "Done" button should perform the same action/function as clicking on the "Cancel" button. As a part of this PR, we've reduxified the loading state of an image as it's easier to work with this afterwards.

## Todo

- [x] Add unit tests for the new action, selectors, reducers;
- [x] Look into the JS errors which pop up after cancelling image editor while image is being loaded (probably the image editor's `canvas` is not removed from DOM after image editor is closed and still expects the image to load)

## Other notes

If we don't pass the `onCancel` prop to the `ImageEditor` component, the "Cancel" button won't render then and the `onCancel` callback becomes a `null`. So, what should happen on clicking the "Done" button in that case? Should we add an option for passing down `onDoneImageNotLoaded` callback?

/cc @gwwar

### CODE REVIEW

**Comment by lancewillett** [comment]
(Sorry for the label changes. Our Waffle.io integration is being weird...)

**Comment by lamosty** [comment]
Ok. I found out that we have a callback issue with `image-editor-canvas` that produces JS errors. Currently, if I close image editor (ie unmount `image-editor-canvas` and other parts), some callback functions will still fire after the image gets loaded.

We need to return early from these functions if the image editor is not visible/opened. I'm considering adding another reducer just for this. Something like `isVisible`.

**Comment by gwwar** [comment]
> So, what should happen on clicking the "Done" button in that case?

Maybe let's decouple the cancel event from which buttons we should render. Do we think the `onDoneImageNotLoaded` handler would differ greatly from what we'd normally do `onCancel`?

**Comment by gwwar** [comment]
> We need to return early from these functions if the image editor is not visible/opened. 

Could we make the callbacks safe to call, if the image editor is not mounted? What js errors are firing?

File: client/blocks/image-editor/image-editor-canvas.jsx [hunk 1]
```diff
@@ -15,9 +15,13 @@ import MediaUtils from 'lib/media/utils';
 import {
 	getImageEditorTransform,
 	getImageEditorFileInfo,
-	getImageEditorCrop
+	getImageEditorCrop,
+	imageEditorIsImageLoaded
```
Comment (gwwar) [inline client/blocks/image-editor/image-editor-canvas.jsx]: Maybe `isImageEditorImageLoaded`? I like putting the get/is at the start so you can tell what the return type is.

**Comment by lamosty** [comment]
> Could we make the callbacks safe to call, if the image editor is not mounted? What js errors are firing?

Yep. I made it safe to call in ccd983f85211ef95b727af1e17f8c3c2252d83ed.

> Maybe let's decouple the cancel event from which buttons we should render.

Even if we do this, if the `onCancel` event handler is not passed in, what should happen after clicking on "Done" if image is not yet loaded? It's the same thing except decoupling those two.

**Comment by lamosty** [comment]
> Do we think the onDoneImageNotLoaded handler would differ greatly from what we'd normally do onCancel?

It depends on the use case. For example, if image editor is used as a step in a flow of setting an image somewhere (e.g. Gravatar), clicking "Done" should take the user to the next step, whether he/she edited the image or not (and even if he/she clicked the "Done" button before the image is loaded). It's not the same as clicking on "Cancel".

File: client/blocks/image-editor/image-editor-canvas.jsx (hunk 1 above)
Comment (lamosty) [inline client/blocks/image-editor/image-editor-canvas.jsx]: Oh, nice idea. Though a bit harder to read. :D

**Comment by gwwar** [comment]
Ah in that case maybe let's pass back an error as the first arg, which the caller should decide what to do in an incomplete case where there is no blob. eg `done( error, data )`

**Comment by lamosty** [comment]
Can I get another review, please? I've fixed/changed the things we talked about @gwwar.

**Comment by gwwar** [comment]
👍 LGTM!

**Comment by lamosty** [comment]
Thanks for the review @gwwar!

**Pull request merged**